### ISSUE TITLE
JACK MIDI support

### ISSUE THREAD

**Issue by simonvanderveldt** [ahlstromcj/sequencer64#45]
First of all thanks a lot for working on seq24/sequencer64, very nice to see it getting some attention and fixes.

Now I'm fairly certain I've read everything in the manual and also couldn't find another relevant issue, but is it correct there's no JACK MIDI support in sequencer64?
Are there plans to add it?

If there's no JACK MIDI support what is the recommended way to run sequencer64 in an otherwise 100% JACK setup?

**Comment by ahlstromcj** [comment]
I would like someday to include full JACK support.  Of course, ALSA MIDI is already working.  There's also PortMIDI support, but I haven't even tested that, and that's more for Windows.  At some point, I will look into providing true JACK MIDI support.

In the meantime, you can use the a2j (ALSA to JACK) MIDI bridge.  Here's what I do, which is also described better in contrib/notes/jack-testing.txt.  Yoshimi is the software synthesizer I use for testing.
Make sure that [manual-alsa=ports

qjackctl &
yoshimi -j -J
a2jmidid --export-hw &

In the QJackCtl MIDI tab, connect the a2j / Midi Through item on the left
to the Yoshimi / midi in item on the right.

sequencer64 -J &

In Live mode, you'll have to arm the desired sequences. Before starting
playback.

Again, see the "Sequencer64 and Yoshimi Tests" section in the jsck-testing file.  I just updated it.

**Comment by simonvanderveldt** [comment]
@ahlstromcj Thanks a lot for the quick reply and the tips. Thus far I've been able to have everything I use work directly with JACK, so no need for a2jmidid or a2jmidi_bridge. Anyway,  it's working now using a2jmidid, which created a pretty big "instrument" in JACK containing 16 MIDI outs and 1 MIDI in in JACK. At least I can get going :)

If adding JACK MIDI support is possible that would be awesome!
Unfortunately I'm not very proficient in C/C++ so can't really help. The only thing I can point to that might be of help is this PR for LMMS which also adds JACK MIDI support on top of the already existing JACK audio support. Maybe there's something in there that can help you.

**Comment by ahlstromcj** [comment]
I'm looking into using either the JACK API, or RtMidi, which provides support for ALSA and JACK (and for Windows and OSX) as well.  Have you heard anything about this library?  It sounds better than PortMidi, which Sequencer64 has inherited from Seq24 (but we only build it, we don't use it, since we don't yet support Windows).

**Comment by simonvanderveldt** [comment]
> I'm looking into using either the JACK API, or RtMidi, which provides support for ALSA and JACK (and for Windows and OSX) as well. Have you heard anything about this library? It sounds better than PortMidi, which Sequencer64 has inherited from Seq24 (but we only build it, we don't use it, since we don't yet support Windows).

Awesome! I don't have experience with rtmidi. PortMidi I have used and even for the simple project I was doing (MIDI note mangling) I had multiple issues with it, so would definitely not recommend it. That was a couple of years ago though.

How multiplatform is sequencer64 supposed to be? Personally I'd gravitate towards a pure JACK midi and audio implementation first and only use another library for ports to non-JACK if they are necessary.

This [SO answer](http://stackoverflow.com/a/9770652) from @rvega + the mailing list thread he links to might me of interest even though it's a couple of years old
http://lists.linuxaudio.org/pipermail/linux-audio-dev/2012-March/031944.html

**Comment by georgkrause** [comment]
@ahlstromcj how is the current state? anything to test yet? ;)

**Comment by ahlstromcj** [comment]
Dang, forgot to close this one.  Native JACK MIDI (with fallback to ALSA) is now the default build in the 0.90.x line in the master branch.  I still have worries about possible I/O starvation, but so far no one has complained.

**Comment by ahlstromcj** [comment]
Closed.  The basic implementation works.  If any issues are found, create a new issue.  Thanks!
